src/centrality/betweenness.c

```diff
@@ -1018,3 +1015,841 @@ int igraph_edge_betweenness_estimate(const igraph_t *graph, igraph_vector_t *res
+ * \param directed Logical, if true directed paths will be considered
+ *        for directed graphs. It is ignored for undirected graphs.
+ * \param weights An optional vector containing edge weights for
+ *        calculating weighted betweenness. No edge weight may be NaN.
+ *        Supply a null pointer here for unweighted betweenness.
+ * \return Error code:
+ *        \c IGRAPH_ENOMEM, not enough memory for
+ *        temporary data.
+ *        \c IGRAPH_EINVVID, invalid vertex id passed in
+ *        \p vids.
+ *
+ * Time complexity: O(|V||E|),
+ * |S| The number of vertices in the group
+ * |E| The number of edges in the graph
+ *
+ * \sa \ref igraph_betweenness() to calculate the exact betweenness and
+ * \ref igraph_edge_betweenness_cutoff() to calculate the range-limited
+ * edge betweenness.
+ */
+int igraph_betweenness_subset(const igraph_t *graph, igraph_vector_t *res,
+                              const igraph_vs_t vids, igraph_bool_t directed,
+                              const igraph_vector_t *weights) {
+
+    long int no_of_nodes = igraph_vcount(graph), to_reach;
+    igraph_dqueue_t q = IGRAPH_DQUEUE_NULL;
+    long int *distance;
+    /* Note: nrgeo holds the number of shortest paths, which may be very large in some cases,
+     * e.g. in a grid graph. If using an integer type, this results in overflow.
+     * With a 'long long int', overflow already affects the result for a grid as small as 36*36.
+     * Therefore, we use a 'double' instead. While a 'double' holds fewer digits than a 'long long int',
+     * i.e. its precision is lower, it is effectively immune to overflow. The impact on the precision
+     * of the final result is negligible. The max betweenness is correct to 14 decimal digits,
+     * i.e. the precision limit of 'double', even for a 101*101 grid graph. */
+    double *nrgeo = 0;
+    double *tmpscore;
+    igraph_stack_t stack = IGRAPH_STACK_NULL;
+    long int source;
+    long int j, nneis;
+    igraph_vector_int_t *neis;
+    igraph_vit_t vit;
+    unsigned char *is_target;
+    igraph_adjlist_t adjlist_out, adjlist_in;
+    igraph_adjlist_t *adjlist_out_p, *adjlist_in_p;
+
+    if (weights) {
+        return igraph_i_betweenness_subset_weighted(graph, res, vids, directed,
+                                                    weights);
+    }
+
+    if (!igraph_vs_is_all(&vids)) {
+        /* subset */
+        IGRAPH_CHECK(igraph_vit_create(graph, vids, &vit));
+        IGRAPH_FINALLY(igraph_vit_destroy, &vit); 
+        to_reach = IGRAPH_VIT_SIZE(vit);
+        igraph_vector_resize(res, no_of_nodes);
+        igraph_vector_fill(res, 0);
+    } else {
+        /* only  */
+        return igraph_betweenness_cutoff(graph, res, vids, directed,
+                                                    weights, -1);
+    }
+
+    is_target = igraph_Calloc(no_of_nodes, unsigned char);
+    if (is_target == 0) {
+        IGRAPH_ERROR("Can't calculate shortest paths", IGRAPH_ENOMEM);
```

Comment:
In all messages, give context about where the message is coming from: `Insufficient memory for subset betweenness.` In all new code, please follow the [error reporting guidelines](https://github.com/igraph/igraph/wiki/Error-reporting-guidelines), so we don't have to spend time to fix it up later.